flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -742,7 +742,7 @@ public void snapshotState(FunctionSnapshotContext context) throws Exception {
 			// case we adjust nextFreeTransactionalId by the range of transactionalIds that could be used for this
 			// scaling up.
 			if (getRuntimeContext().getNumberOfParallelSubtasks() > nextTransactionalIdHint.lastParallelism) {
-				nextFreeTransactionalId += getRuntimeContext().getNumberOfParallelSubtasks() * kafkaProducersPoolSize;
+				nextFreeTransactionalId += (long) getRuntimeContext().getNumberOfParallelSubtasks() * kafkaProducersPoolSize;
```

Comment:
Good change, although that is rather theoretical bug. To trigger it there would need to be more then 1_000_000 subtasks and more then 2000 parallel ongoing checkpoints.